### DISCUSSION TITLE
Static call to instance method

### DISCUSSION THREAD

**Discussion by wilecoyte78** [larastan/larastan#2009] [Q&A]
I am having an issue I am using version larastan 2.9.8 and it says PHPStan - PHP Static Analysis Tool 1.11.7 and I get Static call to instance method App\Classes\ComposeCore::composeLogMessage().

This is a custom class.

**Comment by calebdw** [comment]
So either make the method static or new up an instance

**Comment by wilecoyte78** [comment]
I have tried 

$comp = new ComposeCore();

$comp = $comp::composeLogMessage();

an it doesn't work.

it gives me same error.

**Reply by calebdw** [reply]
```php
$comp = (new ComposeCore())->composeLogMessage();
```

**Reply by calebdw** [reply]
Or just make the method static:

```php
public static function composeLogMessage()
```

**Reply by wilecoyte78** [reply]
The function is already public static function composeLogMessage. I will try the $comp = (new ComposeCore())->composeLogMessage(); on the static function.

**Reply by calebdw** [reply]
Can you show the class with the method?

**Reply by wilecoyte78** [reply]
```
namespace App\Classes;

class ComposeCore
{
    public static function composeLogMessage(string $message, mixed $e = null): string
    {
        $infoString = '';

        // URL.
        try {
            $infoString .= request()->getPathInfo() . (request()->getQueryString() ? ('?' . request()->getQueryString()) : '');
        } catch (Exception $exception) {
            $infoString .= '(unknown URL)';
        }

        if ($e) {
            $trace = array_map(function ($array) {
                return ['file' => $array['file'], 'line' => $array['line'], 'function' => $array['function'], 'class' => $array['class']];
            }, FlattenException::createFromThrowable($e, $e->status ?? 500, [])->getTrace());
            $trace = array_filter($trace, static function ($element) {
                return !preg_match('/logError|composeLogMessage|handleException/', $element['function']);
            });
            $trace = array_values($trace);
            $lastCall = $trace[1];
            $classArr = explode('\\', $lastCall['class']);
            $line = $trace[0]['line'];
            $function = $lastCall['function'];
        } else {
            $trace = array_map(function ($array) {
                $class = $array['class'] ?? '';

                return ['file' => $array['file'], 'line' => $array['line'], 'function' => $array['function'], 'class' => $class];
            }, debug_backtrace());
            $lastCall = $trace[1];
            $classArr = explode('\\', $lastCall['class']);
            $line = $trace[0]['line'];
            $function = $lastCall['function'];
        }

        return '[' . $infoString . '] ' . $classArr[count($classArr) - 1] . '::' . $function . '::' . $line . ': ' . $message;
    }
}
```

**Reply by calebdw** [reply]
Looks fine to me, try clearing the phpstan cache with:


```bash
phpstan clear-result-cache
// or 
phpstan analyze --debug
```

**Reply by wilecoyte78** [reply]
Still not working.

  Line  |  Models/PsNewsletter.php                                                       
 ------ | ------------------------------------------------------------------------------ 
  :83  |  Static call to instance method App\Classes\ComposeCore::composeLogMessage().

**Reply by wilecoyte78** [reply]
PHP runtime version: 8.3.8
PHP version for analysis: 8.3.8 (from runtime)
PHPStan version: 1.11.7

**Reply by calebdw** [reply]
I just copied the file into a project and the static call passed with no issue. There's not much I can do to help, please try to replicate the issue in a new repo with a fresh laravel install. Also try to use the latest PHPStan and Larastan versions

**Reply by wilecoyte78** [reply]
You are correct It works by itself on a new instance. So something in my code is making it throw a false negative. I will just add an igoreErrors rule.

**Reply by calebdw** [reply]
> So something in my code is making it throw a false negative. I will just add an igoreErrors rule.

You might want to narrow down the root cause instead of just ignoring the error...it could be messing with other things as well

**Reply by wilecoyte78** [reply]
That is going to be very difficult because there are 277 files it scans. It will be hard to figure out where it is is coming from because I get 1689 errors of the same type of error to different static methods and 139 files throw that error.

**Reply by calebdw** [reply]
Try running phpstan with `--debug` and `-vvv` for more verbose output

**Reply by calebdw** [reply]
You can also try to keep gradually adding code to the fresh Laravel install until you find out what breaks.

Or you could try a subtractive method where you gradually remove code until it passes

**Reply by wilecoyte78** [reply]
Here is one of the outputs. What am I exactly looking for?

  Line  |  Providers/AppServiceProvider.php                                               
 ------ | ------------------------------------------------------------------------------- 
  :44  |  Static call to instance method App\Classes\SecurityLogWriter::createLogger().  
         🪪  method.staticCall

**Reply by calebdw** [reply]
It's hard to tell exactly without looking at the code myself, but if `App\Classes\SecurityLogWriter::createLogger()` really is a static method then something is wrong with your setup.

Some things to try:
- is something overriding the method definition (like a stub)?
- are you using the [extension-installer](https://github.com/phpstan/extension-installer)?
- what, if any, extensions are you using?
- can you post your phpstan config?

**Reply by wilecoyte78** [reply]
Here is the config:

```
includes:
    - vendor/larastan/larastan/extension.neon

parameters:

    paths:
        - app/

    # Level 9 is the highest level
    level: 5

    ignoreErrors:
        - '#StatefulGuard::isUserAllowed#'

    excludePaths:
        - ./simplesamlphp/*
        - ./app/Classes/FileScan/*
    
    treatPhpDocTypesAsCertain: false
```

**Reply by wilecoyte78** [reply]
I am not using the extension installer.

**Reply by calebdw** [reply]
Your config looks fine, but you shouldn't have to ignore `./simplesamlphp/*` as you're only analyzing `./app` and that's not inside it

**Reply by wilecoyte78** [reply]
I wanted to show an error with a screenshot from VSCode:

 Line  |  Policies/UserPolicy.php                                                                     
 ------ | -------------------------------------------------------------------------------------------- 
  :48  |  Static call to instance method App\Classes\UserCore::currentUserHasPermission().
  
  
![image](https://github.com/user-attachments/assets/dc0439a4-8f3f-42a3-a5ee-a90fc611e3d6)

**Reply by calebdw** [reply]
I don't see the error in the screenshot, and why do you have two definitons?
